shims/spark311/src/main/scala/com/nvidia/spark/rapids/shims/spark311/ParquetCachedBatchSerializer.scala

```diff
@@ -350,61 +360,82 @@ class ParquetCachedBatchSerializer extends CachedBatchSerializer with Arm {
   val BYTES_ALLOWED_PER_BATCH: Long = _2GB - APPROX_PAR_META_DATA
 
   private[rapids] def compressColumnarBatchWithParquet(
-      gpuCB: ColumnarBatch,
-      schema: StructType): List[ParquetCachedBatch] = {
-    val estimatedRowSize = scala.Range(0, gpuCB.numCols()).map { idx =>
-      gpuCB.column(idx).asInstanceOf[GpuColumnVector].getBase.getDeviceMemorySize / gpuCB.numRows()
+      oldGpuCB: ColumnarBatch,
+      schema: StructType,
+      origSchema: StructType): List[ParquetCachedBatch] = {
+    val estimatedRowSize = scala.Range(0, oldGpuCB.numCols()).map { idx =>
+      oldGpuCB.column(idx).asInstanceOf[GpuColumnVector]
+          .getBase.getDeviceMemorySize / oldGpuCB.numRows()
     }.sum
-    val rowsInBatch = (BYTES_ALLOWED_PER_BATCH / estimatedRowSize).toInt
-    val splitIndices = scala.Range(rowsInBatch, gpuCB.numRows(), rowsInBatch)
-    val buffers = new ListBuffer[ParquetCachedBatch]
-    if (splitIndices.nonEmpty) {
-      val splitVectors = new ListBuffer[Array[ColumnVector]]
-      try {
-        for (index <- 0 until gpuCB.numCols()) {
-          splitVectors +=
-              gpuCB.column(index).asInstanceOf[GpuColumnVector].getBase.split(splitIndices: _*)
-        }
 
-        // Splitting the table
-        // e.g. T0 = {col1, col2,...,coln} => split columns into 'm' cols =>
-        // T00= {splitCol1(0), splitCol2(0),...,splitColn(0)}
-        // T01= {splitCol1(1), splitCol2(1),...,splitColn(1)}
-        // ...
-        // T0m= {splitCol1(m), splitCol2(m),...,splitColn(m)}
-        def makeTableForIndex(i: Int): Table = {
-          val columns = splitVectors.indices.map(j => splitVectors(j)(i))
-          new Table(columns: _*)
+    val columns = for (i <- 0 until oldGpuCB.numCols()) yield {
+      val gpuVector = oldGpuCB.column(i).asInstanceOf[GpuColumnVector]
+      var dataType = origSchema(i).dataType
+      val v = ColumnUtil.ifTrueThenDeepConvertTypeAtoTypeB(gpuVector.getBase,
```

Comment:
Just like in the other file I think it would be a lot cleaner to modify the Spark schema to match what you want and then have a high level API that knows how to make them match with bit casting, instead of trying to make the API overly generic.